src/Grpc.AspNetCore.Server/Internal/GrpcEventSource.cs

```diff
@@ -0,0 +1,159 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System.Diagnostics.Tracing;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Grpc.Core;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal class GrpcEventSource : EventSource
+    {
+        public static readonly GrpcEventSource Log = new GrpcEventSource();
+
+        private PollingCounter? _totalCallsCounter;
+        private PollingCounter? _currentCallsCounter;
+        private PollingCounter? _messagesSentCounter;
+        private PollingCounter? _messagesReceivedCounter;
+        private PollingCounter? _callsFailedCounter;
+        private PollingCounter? _callsDeadlineExceededCounter;
+        private PollingCounter? _callsUnimplementedCounter;
+
+        private long _totalCalls;
+        private long _currentCalls;
+        private long _messageSent;
+        private long _messageReceived;
+        private long _callsFailed;
+        private long _callsDeadlineExceeded;
+        private long _callsUnimplemented;
+
+        internal GrpcEventSource()
+            : base("Grpc.AspNetCore.Server")
+        {
+        }
+
+        // Used for testing
+        internal GrpcEventSource(string eventSourceName)
+            : base(eventSourceName)
+        {
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        [Event(eventId: 1, Level = EventLevel.Verbose)]
+        public void CallStart(string method)
+        {
+            Interlocked.Increment(ref _totalCalls);
+            Interlocked.Increment(ref _currentCalls);
+
+            WriteEvent(1, method);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        [Event(eventId: 2, Level = EventLevel.Verbose)]
+        public void CallStop()
+        {
+            Interlocked.Decrement(ref _currentCalls);
+
+            WriteEvent(2);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        [Event(eventId: 3, Level = EventLevel.Error)]
+        public void CallFailed(StatusCode statusCode)
+        {
+            Interlocked.Increment(ref _callsFailed);
+
+            WriteEvent(3, (int)statusCode);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        [Event(eventId: 4, Level = EventLevel.Error)]
+        public void CallDeadlineExceeded()
+        {
+            Interlocked.Increment(ref _callsDeadlineExceeded);
```

Comment:
Should this increase the number of callsFailed?

Reply:
It will be increased when the call ends.